File: asterix-rs/tests/test_cat025.rs
//! Integration test for CAT025 parsing
//! CNS/ATM Ground System Status Reports
//!
//! This test verifies the Rust API for parsing ASTERIX Category 025 data.
//!
//! Requirements Coverage:
//! - REQ-HLR-001: Parse ASTERIX binary data
//! - REQ-HLR-SYS-001: Parse ASTERIX categories
//! - REQ-LLR-025-*: Additional requirements from Low_Level_Requirements_CAT025.md
//!
//! Test Cases:
//! - TC-RS-CAT025-001: Parse CAT025 binary data
//! - TC-RS-CAT025-002: Verify parsed data structure
//! - TC-RS-CAT025-003: Test error handling
//! - TC-RS-CAT025-004: Test API usage

use asterix::{parse, AsterixError, ParseOptions};

use std::path::PathBuf;

/// Helper function to get path to sample data
#[allow(dead_code)]
fn sample_data_path(filename: &str) -> PathBuf {
    let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.pop(); // Go up from asterix-rs/
    path.push("install/sample_data");
    path.push(filename);
    path
}

#[test]
fn test_parse_cat025_packet() {
    //! Test parsing CAT025 packet
    //!
    //! Requirement: REQ-HLR-001, REQ-LLR-025-010
    //! Verification: TC-RS-CAT025-001

    // Create minimal valid CAT025 packet
    // Structure: [Category: 25] [Length MSB: 0x00] [Length LSB: 0x03]
    let cat025_packet = vec![
        25,   // Category 25
        0x00, // Length MSB
        0x03, // Length LSB = 3 bytes (header only)
    ];

    let options = ParseOptions {
        verbose: false,
        filter_category: None,
        max_records: None,
    };

    let result = parse(&cat025_packet, options);

    // Verify parsing succeeded or handled gracefully
    match result {
        Ok(records) => {
            // If packet is too minimal, it may return empty list, which is acceptable
            if !records.is_empty() {
                assert_eq!(records[0].category, 25, "Should identify as CAT025");
                println!("✓ Parsed CAT025 packet: {} record(s)", records.len());
            }
        }
        Err(e) => {
            // Parsing minimal packet may fail, which is acceptable
            println!("⚠ CAT025 minimal packet parsing: {e:?}");
        }
    }
}

#[test]
fn test_parse_cat025_with_data_items() {
    //! Test parsing CAT025 with data items
    //!
    //! Requirement: REQ-LLR-025-010
    //! Verification: TC-RS-CAT025-002

    // Create CAT025 packet with I025/010 (Data Source Identifier)
    // Structure: [Category: 25] [Length MSB: 0x00] [Length LSB: 0x08] [FSPEC: 0x80] [I025/010: 2 bytes]
    let cat025_packet = vec![
        25,   // Category 25
        0x00, // Length MSB
        0x08, // Length LSB = 8 bytes total
        0x80, // FSPEC: I025/010 present (bit 7 set)
        0x01, // I025/010: SAC = 0x01
        0x23, // I025/010: SIC = 0x23
    ];

    let options = ParseOptions {
        verbose: false,
        filter_category: None,
        max_records: None,
    };

    let result = parse(&cat025_packet, options);

    match result {
        Ok(records) => {
            if !records.is_empty() {
                assert_eq!(records[0].category, 25, "Should identify as CAT025");
                assert!(records[0].length > 0, "Record length should be positive");
                println!(
                    "✓ Parsed CAT025 with data items: {} record(s)",
                    records.len()
                );
            }
        }
        Err(e) => {
            // May fail if configuration not loaded, which is acceptable
            println!("⚠ CAT025 with data items: {e:?}");
        }
    }
}

#[test]
fn test_cat025_error_handling() {
    //! Test error handling for invalid CAT025 data
    //!
    //! Requirement: REQ-HLR-001 (Error handling)
    //! Verification: TC-RS-CAT025-003

    // Test with empty data
    let empty_data = b"";
    let result = parse(empty_data, ParseOptions::default());
    assert!(result.is_err(), "Expected error for empty data");
    println!("✓ Empty data correctly rejected");

    // Test with invalid category (wrong category byte)
    let invalid_packet = vec![
        0xFF, // Invalid category
        0x00, // Length MSB
        0x03, // Length LSB
    ];

    let result = parse(&invalid_packet, ParseOptions::default());
    // Should handle gracefully (return error or empty list)
    match result {
        Ok(records) => {
            // May return empty list for invalid category, which is acceptable
            println!("✓ Invalid category handled: {} records", records.len());
        }
        Err(AsterixError::InvalidCategory { category, .. }) => {
            assert_eq!(category, 255);
            println!("✓ Correctly rejected invalid category {category}");
        }
        Err(e) => {
            println!("✓ Rejected invalid data: {e:?}");
        }
    }
}

#[test]
fn test_cat025_api_usage() {
    //! Test Rust API usage for CAT025
    //!
    //! Requirement: REQ-HLR-001
    //! Verification: TC-RS-CAT025-004

    // Create minimal valid packet
    let cat025_packet = vec![
        25,   // Category 25
        0x00, // Length MSB
        0x03, // Length LSB = 3 bytes
    ];

    // Test with verbose mode
    let options_verbose = ParseOptions {
        verbose: true,
        filter_category: None,
        max_records: None,
    };

    let options_quiet = ParseOptions {
        verbose: false,
        filter_category: None,
        max_records: None,
    };

    let result_verbose = parse(&cat025_packet, options_verbose);
    let result_quiet = parse(&cat025_packet, options_quiet);

    // Both should handle gracefully
    match (result_verbose, result_quiet) {
        (Ok(records_v), Ok(records_q)) => {
            assert_eq!(
                records_v.len(),
                records_q.len(),
                "Both modes should return same number of records"
            );
            println!(
                "✓ CAT025 API usage: verbose={}, quiet={} records",
                records_v.len(),
                records_q.len()
            );
        }
        (Ok(records_v), Err(_)) => {
            println!(
                "⚠ CAT025 verbose succeeded ({} records) but quiet failed",
                records_v.len()
            );
        }
        (Err(_), Ok(records_q)) => {
            println!(
                "⚠ CAT025 quiet succeeded ({} records) but verbose failed",
                records_q.len()
            );
        }
        (Err(e1), Err(e2)) => {
            // Both failed, which may be acceptable for minimal packets
            println!("⚠ CAT025 both modes failed: {e1:?} / {e2:?}");
        }
    }

    // Test with category filter
    let options_filter = ParseOptions {
        verbose: false,
        filter_category: Some(25),
        max_records: None,
    };

    let result_filter = parse(&cat025_packet, options_filter);
    match result_filter {
        Ok(records) => {
            for record in &records {
                assert_eq!(record.category, 25, "Filtered records should be CAT025");
            }
            println!(
                "✓ Category filter working: {} CAT025 records",
                records.len()
            );
        }
        Err(e) => {
            println!("⚠ Category filter test: {e:?}");
        }
    }
}

File: asterix/test/test_cat025.py
"""
Python integration test for CAT025 parsing
CNS/ATM Ground System Status Reports

This test verifies the Python API for parsing ASTERIX Category 025 data.

Requirements Coverage:
- REQ-HLR-001: Parse ASTERIX binary data
- REQ-HLR-SYS-001: Parse ASTERIX categories
- REQ-LLR-025-*: Additional requirements from Low_Level_Requirements_CAT025.md

Test Cases:
- TC-PY-CAT025-001: Parse CAT025 binary data
- TC-PY-CAT025-002: Verify parsed data structure
- TC-PY-CAT025-003: Test error handling
"""

import pytest
import asterix


class TestCAT025:
    """Test Category 025 (CNS/ATM Ground System Status Reports) parsing"""

    def test_parse_cat025_packet(self):
        """
        Test parsing CAT025 packet

        Requirement: REQ-HLR-001, REQ-LLR-025-010
        Verification: TC-PY-CAT025-001
        """
        # Create minimal valid CAT025 packet
        # Structure: [Category: 0x19] [Length MSB: 0x00] [Length LSB: 0x03]
        cat025_packet = bytearray([
            0x19,  # Category 25
            0x00,  # Length MSB
            0x03,  # Length LSB = 3 bytes (header only)
        ])

        result = asterix.parse(cat025_packet)

        # Verify parsing succeeded
        assert result is not None, "Parse result should not be None"
        assert isinstance(result, list), "Parse should return a list"
        
        # If packet is too minimal, it may return empty list, which is acceptable
        if len(result) > 0:
            assert result[0]['category'] == 25, "Should identify as CAT025"

    def test_parse_cat025_with_data_items(self):
        """
        Test parsing CAT025 with data items

        Requirement: REQ-LLR-025-010
        Verification: TC-PY-CAT025-002
        """
        # Create CAT025 packet with I025/010 (Data Source Identifier)
        # Structure: [Category: 0x19] [Length MSB: 0x00] [Length LSB: 0x08] [FSPEC: 0x80] [I025/010: 2 bytes]
        cat025_packet = bytearray([
            0x19,  # Category 25
            0x00,  # Length MSB
            0x08,  # Length LSB = 8 bytes total
            0x80,  # FSPEC: I025/010 present (bit 7 set)
            0x01,  # I025/010: SAC = 0x01
            0x23,  # I025/010: SIC = 0x23
        ])

        result = asterix.parse(cat025_packet)

        # Verify parsing succeeded
        assert result is not None, "Parse result should not be None"
        assert isinstance(result, list), "Parse should return a list"
        
        if len(result) > 0:
            assert result[0]['category'] == 25, "Should identify as CAT025"

    def test_cat025_error_handling(self):
        """
        Test error handling for invalid CAT025 data

        Requirement: REQ-HLR-001 (Error handling)
        Verification: TC-PY-CAT025-003
        """
        # Test with empty data
        # Security fix: Empty data should raise ValueError (not return empty list)
        empty_data = bytearray([])
        with pytest.raises(ValueError, match="Empty input data"):
            asterix.parse(empty_data)


        # Test with invalid category (wrong category byte)
        invalid_packet = bytearray([
            0xFF,  # Invalid category
            0x00,  # Length MSB
            0x03,  # Length LSB
        ])
        result = asterix.parse(invalid_packet)
        # Should handle gracefully (return empty list or valid list with error indication)
        assert isinstance(result, list), "Should return list even for invalid data"

    def test_cat025_api_usage(self):
        """
        Test Python API usage for CAT025

        Requirement: REQ-HLR-001
        Verification: TC-PY-CAT025-004
        """
        # Create minimal valid packet
        cat025_packet = bytearray([
            0x19,  # Category 25
            0x00,  # Length MSB
            0x03,  # Length LSB = 3 bytes
        ])

        # Test with verbose mode
        result_verbose = asterix.parse(cat025_packet, verbose=True)
        result_quiet = asterix.parse(cat025_packet, verbose=False)

        # Both should return valid results
        assert isinstance(result_verbose, list), "Verbose mode should return list"
        assert isinstance(result_quiet, list), "Quiet mode should return list"
        assert len(result_verbose) == len(result_quiet), "Both modes should return same number of records"
